### PR TITLE
KAA-177: Implement basic logger for C SDK.

### PR DESCRIPTION
**Pull request by ghost**
Reviewers - Andrey Panasenko and Taras Lemkin.

### CODE REVIEW

File: client/client-multi/client-c/src/kaa_log.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#include <stdarg.h>
+#include <string.h>
+
+#include "kaa_common.h"
+
+// TODO: Need to print milliseconds
+#define KAA_LOG_PREFIX_FORMAT    "%04d/%02d/%02d %d:%02d:%02d [%s] [%s:%d] (%d) - "
+
+static char* kaa_log_level_name[] = {
+          "OFF"
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: "NONE" is better than "OFF"

File: client/client-multi/client-c/src/kaa_log.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+static const size_t log_buffer_size= sizeof(log_buffer) / sizeof(char);
+
+static void write_log(char* log, size_t len);
+
+void kaa_log_init(kaa_log_level_t level, FILE* sink) {
+    max_log_level = level;
+    kaa_set_log_sync(sink);
+}
+
+void kaa_log_deinit() {
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: Please move "{" to a new line here and everywhere else.

File: client/client-multi/client-c/src/kaa_log.c [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    res_len = vsnprintf(log_buffer + consumed_len
+                      , log_buffer_size - consumed_len
+                      , format
+                      , args);
+    va_end(args);
+
+    if (res_len > 0) { consumed_len += res_len; }
+
+    if (consumed_len > 0) {
+        write_log(log_buffer ,consumed_len);
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: write_log(log_buffer, consumed_len);

File: client/client-multi/client-c/src/kaa_log.c [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    // TODO: Need to print milliseconds. May be used timespec() from C11 standard,
+    // but GCC 4.6.4 doesn't support this API.
+
+    // Load data into a buffer
+    res_len = snprintf(log_buffer, log_buffer_size, KAA_LOG_PREFIX_FORMAT,
+                        1900 + tp->tm_year, tp->tm_mon + 1, tp->tm_mday,
+                        tp->tm_hour, tp->tm_min, tp->tm_sec,
+                        kaa_log_level_name[level], trancated_name, lineno, error_code);
+
+    if (res_len > 0) { consumed_len += res_len; }
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: Please move statement to a new line

File: client/client-multi/client-c/src/kaa_log.c [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    va_list args;
+    va_start(args, format);
+    res_len = vsnprintf(log_buffer + consumed_len
+                      , log_buffer_size - consumed_len
+                      , format
+                      , args);
+    va_end(args);
+
+    if (res_len > 0) { consumed_len += res_len; }
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: Please move statement to a new line

File: client/client-multi/client-c/src/kaa_log.h [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * the message will be truncated to match the @link KAA_MAX_LOG_MESSAGE_LENGTH @endlink .</p>
+ *
+ */
+#define KAA_MAX_LOG_MESSAGE_LENGTH   512
+
+typedef enum kaa_log_level_t {
+    /**
+     * Use to switch off logging.
+     */
+    KAA_LOG_OFF,
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: KAA_LOG_NONE is better

File: client/client-multi/client-c/src/kaa_log.h [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+     */
+    KAA_LOG_FATAL,
+    KAA_LOG_ERROR,
+    KAA_LOG_WARN,
+    KAA_LOG_INFO,
+    KAA_LOG_DEBUG,
+    KAA_LOG_TRACE,
+
+    /**
+     * <p>Use for service purpose.</p>
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Used

File: client/client-multi/client-c/src/kaa_log.h [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+ * @link KAA_LOG_INFO @endlink, @link KAA_LOG_DEBUG @endlink,
+ * @link KAA_LOG_TRACE @endlink values can be used.</p>
+ *
+ * <p>In order to switch off logging use @link KAA_LOG_OFF @endlink .</p>
+ *
+ * <p><If <i>sink</i> parameter is set to <i>NULL</i>, standard output
+ * (<i>stdout</i>) will be used.</p>
+ *
+ * @param level Max log level to be used.
+ * @param sink Place where logs to be put.
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: @param sink Destination file for the logs.

File: client/client-multi/client-c/src/kaa_log.h [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+ * <p>In order to switch off logging use @link KAA_LOG_OFF @endlink .</p>
+ *
+ * @param new_log_level Max log level to be used.
+ *
+ * @see kaa_log_level_t
+ */
+void kaa_set_log_level(kaa_log_level_t new_log_level);
+
+/**
+ * <p>Form log message and put it the sink.</p>
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Formats the log message and puts into into the sink.

File: client/client-multi/client-c/src/kaa.c [hunk 10]
```diff
@@ -74,6 +79,7 @@ kaa_error_t kaa_init()
 
 kaa_error_t kaa_deinit()
 {
+    kaa_log_deinit();
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa.c]: Deinitialize logger after context is destroyed.

File: client/client-multi/client-c/src/kaa_log.c [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        return;
+    }
+
+    int res_len = 0;
+    size_t consumed_len = 0;
+
+    // Truncate full file name
+    char* path_separator_pos = strrchr(filename, '/');
+    path_separator_pos = (path_separator_pos ? path_separator_pos : strrchr(filename, '\\'));
+    const char* trancated_name = (path_separator_pos ? path_separator_pos + 1 : filename);
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: truncated_name

File: client/client-multi/client-c/src/kaa_log.c [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    struct tm* tp = gmtime(&t);
+
+    // TODO: Need to print milliseconds. May be used timespec() from C11 standard,
+    // but GCC 4.6.4 doesn't support this API.
+
+    // Load data into a buffer
+    res_len = snprintf(log_buffer, log_buffer_size, KAA_LOG_PREFIX_FORMAT,
+                        1900 + tp->tm_year, tp->tm_mon + 1, tp->tm_mday,
+                        tp->tm_hour, tp->tm_min, tp->tm_sec,
+                        kaa_log_level_name[level], trancated_name, lineno, error_code);
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: truncated_name

File: client/client-multi/client-c/src/kaa_log.c [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+                            , kaa_error_t error_code, const char* format, ...)
+{
+    if (level > max_log_level) {
+        return;
+    }
+
+    int res_len = 0;
+    size_t consumed_len = 0;
+
+    // Truncate full file name
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: Truncate the file name

File: client/client-multi/client-c/src/kaa_log.c [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    char* path_separator_pos = strrchr(filename, '/');
+    path_separator_pos = (path_separator_pos ? path_separator_pos : strrchr(filename, '\\'));
+    const char* trancated_name = (path_separator_pos ? path_separator_pos + 1 : filename);
+
+    // Convert to UTC time
+    time_t t = time(NULL);
+    struct tm* tp = gmtime(&t);
+
+    // TODO: Need to print milliseconds. May be used timespec() from C11 standard,
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: TODO: Need to print milliseconds. For this purpose, timespec() from C11 standard may be used,

File: client/client-multi/client-c/src/kaa_log.c [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    // but GCC 4.6.4 doesn't support this API.
+
+    // Load data into a buffer
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.c]: Load data into the buffer

File: client/client-multi/client-c/src/kaa_log.h [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+#define KAA_MAX_LOG_MESSAGE_LENGTH   512
+
+typedef enum kaa_log_level_t {
+    /**
+     * Use to switch off logging.
+     */
+    KAA_LOG_OFF,
+
+    /**
+     * Use to configure log level.
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Use to configure the log level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+     * <p>Use for service purpose.</p>
+     *
+     * <p><b>NOTE:</b> Do not use these values directly.</p>
+     */
+    KAA_LOG_LEVEL_MIN = KAA_LOG_OFF,
+    KAA_LOG_LEVEL_MAX = KAA_LOG_TRACE
+} kaa_log_level_t;
+
+/**
+ * <p>Initialize logger.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: <p>Initialize the logger.</p>

File: client/client-multi/client-c/src/kaa_log.h [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p>To configure log level, one of @link KAA_LOG_FATAL @endlink,
+ * @link KAA_LOG_ERROR @endlink, @link KAA_LOG_WARN @endlink,
+ * @link KAA_LOG_INFO @endlink, @link KAA_LOG_DEBUG @endlink,
+ * @link KAA_LOG_TRACE @endlink values can be used.</p>
+ *
+ * <p>In order to switch off logging use @link KAA_LOG_OFF @endlink .</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: <p>To switch off logging, use @link KAA_LOG_OFF @endlink .</p>

File: client/client-multi/client-c/src/kaa_log.h [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p><If <i>sink</i> parameter is set to <i>NULL</i>, standard output
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h:75]: If the <i>sink</i> parameter is set to <i>NULL</i>, standard output

File: client/client-multi/client-c/src/kaa_log.h [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * @see kaa_log_level_t
+ *
+ */
+void kaa_log_init(kaa_log_level_t level, FILE* sink);
+
+/**
+ * <p>Deinitialize Kaa logger.</p>
+ *
+ * <p>Close user sink, if it was used.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Close the user sink, if it was used.

File: client/client-multi/client-c/src/kaa_log.h [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ */
+void kaa_log_deinit();
+
+/**
+ * <p>Set sink where logs to be put.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Set the sink for the logs.

File: client/client-multi/client-c/src/kaa_log.h [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p><If <i>sink</i> parameter is set to <i>NULL</i>, standard output
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: If the ...

File: client/client-multi/client-c/src/kaa_log.h [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+ * (<i>stdout</i>) will be used.</p>
+ *
+ */
+void kaa_set_log_sync(FILE* sink);
+
+/**
+ * <p>Retrieve current log level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Retrieve the current log level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+ */
+kaa_log_level_t kaa_get_log_level();
+
+/**
+ * <p>Set log level.</p>
+ *
+ * <p>To configure log level, one of @link KAA_LOG_FATAL @endlink,
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: To configure the log level, ...

File: client/client-multi/client-c/src/kaa_log.h [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+ */
+void kaa_set_log_sync(FILE* sink);
+
+/**
+ * <p>Retrieve current log level.</p>
+ */
+kaa_log_level_t kaa_get_log_level();
+
+/**
+ * <p>Set log level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: <p>Set the log level.</p>

File: client/client-multi/client-c/src/kaa_log.h [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p>To configure log level, one of @link KAA_LOG_FATAL @endlink,
+ * @link KAA_LOG_ERROR @endlink, @link KAA_LOG_WARN @endlink,
+ * @link KAA_LOG_INFO @endlink, @link KAA_LOG_DEBUG @endlink,
+ * @link KAA_LOG_TRACE @endlink values can be used.</p>
+ *
+ * <p>In order to switch off logging use @link KAA_LOG_OFF @endlink .</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: To switch off logging, use ...

File: client/client-multi/client-c/src/kaa_log.h (hunk 9 above)
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: Form a log message and put it in the sink.

File: client/client-multi/client-c/src/kaa_log.h [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+ * @param new_log_level Max log level to be used.
+ *
+ * @see kaa_log_level_t
+ */
+void kaa_set_log_level(kaa_log_level_t new_log_level);
+
+/**
+ * <p>Form log message and put it the sink.</p>
+ *
+ * <p>Message format is following:
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The message format is the following:

File: client/client-multi/client-c/src/kaa_log.h [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+ * @see KAA_MAX_LOG_MESSAGE_LENGTH
+ * @see kaa_log_level_t
+ * @see kaa_error_t
+ *
+ */
+void kaa_log_write(const char* filename, int lineno, kaa_log_level_t level
+                            , kaa_error_t error_code, const char* format, ...);
+
+/**
+ * <p>Log message with @link KAA_LOG_FATAL @endlink level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The log message with the @link KAA_LOG_FATAL @endlink level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+ * <p>Log message with @link KAA_LOG_ERROR @endlink level.</p>
+ *
+ * <p>For more details see @link kaa_log_write @endlink .</p>
+ *
+ * @see kaa_error_t
+ */
+#define KAA_LOG_ERROR(err, ...) kaa_log_write(__FILE__, __LINE__, KAA_LOG_ERROR, err, __VA_ARGS__);
+
+/**
+ * <p>Log message with @link KAA_LOG_WARN @endlink level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The log message with the @link KAA_LOG_WARN @endlink level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p>For more details see @link kaa_log_write @endlink .</p>
+ *
+ * @see kaa_error_t
+ */
+#define KAA_LOG_WARN(err, ...)  kaa_log_write(__FILE__, __LINE__, KAA_LOG_WARN, err, __VA_ARGS__);
+
+/**
+ * <p>Log message with @link KAA_LOG_INFO @endlink level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The log message with the @link KAA_LOG_INFO @endlink level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 31]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p>For more details see @link kaa_log_write @endlink .</p>
+ *
+ * @see kaa_error_t
+ */
+#define KAA_LOG_INFO(err, ...)  kaa_log_write(__FILE__, __LINE__, KAA_LOG_INFO, err, __VA_ARGS__);
+
+/**
+ * <p>Log message with @link KAA_LOG_DEBUG @endlink level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The log message with the @link KAA_LOG_DEBUG @endlink level.

File: client/client-multi/client-c/src/kaa_log.h [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * <p>For more details see @link kaa_log_write @endlink .</p>
+ *
+ * @see kaa_error_t
+ */
+#define KAA_LOG_DEBUG(err, ...) kaa_log_write(__FILE__, __LINE__, KAA_LOG_DEBUG, err, __VA_ARGS__);
+
+/**
+ * <p>Log message with @link KAA_LOG_TRACE @endlink level.</p>
```
Comment (vzhukovskyi) [inline client/client-multi/client-c/src/kaa_log.h]: The log message with the @link KAA_LOG_TRACE @endlink level.

**Pull request merged**